src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -254,164 +255,314 @@ public void SetCapacityIfLarger(int value)
     }
 
     public void Add(T item)
+        => Insert(Count, item);
+
+    public void AddRange(ImmutableArray<T> items)
+        => InsertRange(Count, items);
+
+    public void AddRange(ReadOnlySpan<T> items)
     {
+        // Note: We don't delegate this overload to InsertRange(ReadOnlySpan<T>) because
+        // ImmutableArray<T>.Builder supports an AddRange overload that takes a ReadOnlySpan<T>.
+        // Delegating to InsertRange(ReadOnlySpan<T>) could unnecessarily introduce an
+        // array allocation and copy.
+
+        if (items.IsEmpty)
+        {
+            return;
+        }
+
         if (TryGetBuilderAndEnsureCapacity(out var builder))
         {
-            builder.Add(item);
+            builder.AddRange(items);
         }
-        else if (_inlineCount < InlineCapacity)
+        else if (_inlineCount + items.Length <= InlineCapacity)
         {
-            SetInlineElement(_inlineCount, item);
-            _inlineCount++;
+            foreach (var item in items)
+            {
+                SetInlineElement(_inlineCount, item);
+                _inlineCount++;
+            }
         }
         else
         {
-            Debug.Assert(_inlineCount == InlineCapacity);
             MoveInlineItemsToBuilder();
-            _builder.Add(item);
+            _builder.AddRange(items);
         }
     }
 
-    public void AddRange(ImmutableArray<T> items)
+    public void AddRange<TList>(TList list)
+        where TList : struct, IReadOnlyList<T>
+        => AddRange(list, startIndex: 0, list.Count);
+
+    public void AddRange<TList>(TList list, int startIndex, int count)
+        where TList : struct, IReadOnlyList<T>
     {
-        AddRange(items.AsSpan());
+        // Note: We don't delegate this overload to InsertRange<TList>(TList, int, int) because
+        // it requires extra allocations that aren't necessary for AddRange(...).
+
+        if (count == 0)
+        {
+            return;
+        }
+
+        var (start, end) = (startIndex, startIndex + count - 1);
+
+        if (TryGetBuilderAndEnsureCapacity(out var builder))
```

Comment:
>TryGetBuilderAndEnsureCapacity [](http://example.com/codeflow?start=12&length=30)

I find it a little confusing around when _builder can be used in this class. This class doesn't support concurrency, so I'm not sure why TryGetBuilderAndEnsureCapacity returns a builder.

Reply:
I didn't quite understand your comment. Do you have a question I can try and answer?

Reply:
I find it odd that a method like this uses both a local builder and the member _builder when they are the same object. Is there a reason why the TryGetBuilder* methods return _builder if concurrency isn't a concern?

Reply:
Do you mean that you'd prefer that we just always access `_builder`? Or, would you rather than `MoveInlineitemsToBuilder()` also provide a local for the `_builder`. I had started the latter refactoring but decided that it was a bit much for this pull request.

Are you suggesting a change before signing off? I'm happy to discuss further refactoring of this struct and better alignment with Roslyn's `TemporaryArray<T>`.